challenge_find_my_mates/src/find_my_mates.py

```diff
@@ -0,0 +1,338 @@
+
+
+class ReportPeople(State):
+    """
+    Form the sentences for all three people and say that sentence
+    """
+
+    def __init__(self, robot):
+        State.__init__(self, outcomes=['Done', 'Aborted', 'Failed'])
+        self._robot = robot
+        self._room_entity = self._robot.ed.get_entity(id=ROOM_ID)
+
+    def execute(self, userdata=None):
+        entities = self._robot.ed.get_entities()
+        person_entities = [entity for entity in entities if entity.is_a("person")]
+
+        furniture_entities = [entity for entity in entities if entity.is_a("furniture")]
+
+        # Match the furniture entities to rooms
+        furniture_entities_room = []
+        for item in furniture_entities:
+            if self._room_entity.in_volume(kdl_conversions.VectorStamped(vector=item._pose.p), "in"):
+                furniture_entities_room.append(item)
+        closest_entity = []
+        for person in person_entities:
+            distance_closest = 10.0
+            temp_entity_near = furniture_entities_room[0]
+            for entity in furniture_entities_room:
+                distance = entity.distance_to_2d(person.position)
+                if distance < distance_closest:
+                    temp_entity_near = entity
+                    distance_closest = distance
+            closest_entity.append(temp_entity_near)
+
+        self._robot.speech.speak('I found many people but none were as pretty as you.', block=True)
+        people_descriptions = {}
+        for i, person in enumerate(person_entities):
+            sentence = "I found {} near the {}.\n".format(getattr(person, 'name'), closest_entity[i].id)
+
+            unique_property = None
+            properties = [prop for prop in dir(person) if not prop.startswith('__')]
+            attributes = dict()
+            for prop in properties:
+                temp_prop = [getattr(person, prop) for person in person_entities]
+                c = Counter(temp_prop)
+                attributes[prop] = c[getattr(person, prop)]
+                if prop == 'gender_confidence':
+                    if any(j <= 68 for j in temp_prop):
+                        del attributes['gender']
+                    del attributes['gender_confidence']
+                del attributes['id']
+                for attr in attributes:
+                    if attributes[attr] == 1:
+                        unique_property = attr
+
+            # Edit these sentences based on available attributes
+            if unique_property == 'gender':
+                sentence += "and the person was a {}.".format(getattr(person, unique_property))
+            elif unique_property == 'tall':
+                sentence += "and the person was tall. "
+            elif unique_property == 'shirt_colour':
+                sentence += "and the person was wearing a mostly {} shirt.".format(getattr(person, unique_property))
+
+            people_descriptions[getattr(person, 'name')] = sentence
+            if i > 3:
```

Comment:
Why not create a unique description for all people? Or are we certain that the operator will ask for descriptions of the first three people?

Reply:
This is all people, this if is only met when there are four people (```>``` not ```=>```), and the challenge states there will be four people. But that should then also be the end of the loop so the ```break``` is unnecessary.